llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -3743,7 +3743,7 @@ InstructionCost AArch64TTIImpl::getInterleavedMemoryOpCost(
   assert(Factor >= 2 && "Invalid interleave factor");
   auto *VecVTy = cast<VectorType>(VecTy);
 
-  if (VecTy->isScalableTy() && (!ST->hasSVE() || Factor != 2))
+  if (VecTy->isScalableTy() && !ST->hasSVE())
```

Comment:
should this assert that Factor == 2 for SVE to guard against accidentally returning an incorrect non-invalid cost? Same for RISCV

Reply:
From what I can tell, the existing costs should be valid for Factor > 2.  We already use them for fixed vector interleaves and for VF=2 scalable, so the code should be plausible.  Once we add an IR representation for these (which we don't currently have), the costs will need audited, but the current implementation seems sane.

You seem reasonable opposed to what I consider a simple cleanup.  I'm not particularly motivated to keep pushing this, if you'd rather I just drop the patch, please say so.

Reply:
If I understand correctly, I think all @fhahn is worried about here is a situation where the vectoriser may support larger interleave factors for scalable vectors but relies on the cost model returning Invalid for factors they don't support. So having an assert that Factor == 2 with scalable vectors for both RISC and AArch64 seems reasonable I guess?

The intent of this patch seems reasonable to me - by bailing out early in interleavedAccessCanBeWidened we can avoid doing unnecessary work. It's perhaps worth pointing that there is currently a patch (https://github.com/llvm/llvm-project/pull/89018) to add support for factors > 2.

Reply:
I'm with @preames, there's no legal LLVM IR that cannot be handled by the code generator when scalable vectors are available and thus there's no reason for the cost model to return an invalid cost. https://github.com/llvm/llvm-project/pull/89018 changes nothing because it doesn't introduce any new IR and thus at worst the cost might be wrong, but it'll still safely make it through code generation.

The cost model shouldn't have to worry about non-existent LLVM IR, whereas it seems prudent for LoopVectorize to first check that what it's trying to do is possible within the current limitations of LLVM IR.